### PR TITLE
Select All Feature

### PR DESCRIPTION
**Pull request by diogogmatos**
@Carlos-oficial Implemented a Select All feature that allows the users to select all courses in a semester for each semester of each year. I contributed with the styling and design and also made improvements and bug fixes, like adding compatibility for saving the `Select All` checkboxes on local storage to remember the user's choices.

Also added some UMinho events, namely the start and end of each semester, according to the official 2022/2023 calendar from UMinho. 

**Preview:**
image

### CODE REVIEW

**Review by MarioRodrigues10 — DISMISSED** [review]
Please run formatter and linter :pray:

File: components/CheckBox/CheckBox.tsx [hunk 1]
```diff
@@ -8,19 +8,25 @@ const { Panel } = Collapse;
...
 function CheckBox({ filters, handleFilters }) {
   const [Checked, setChecked] = React.useState<number[]>([]);
+
   React.useEffect(() => {
     const stored: number[] = JSON.parse(localStorage.getItem("checked")) ?? [];
     setChecked(stored);
     handleFilters(stored);
   }, []);
 
+  React.useEffect(() => {
```
Comment (EnzoVieira) [inline components/CheckBox/CheckBox.tsx:18]: I'd suggest some changes, like:
for what is this useEffect being used for?

if it is not used, you can delete it.

File: components/CheckBox/CheckBox.tsx [hunk 2]
```diff
@@ -65,15 +71,45 @@ function CheckBox({ filters, handleFilters }) {
...
+
     setChecked(newCheck);
     localStorage.setItem("checked", JSON.stringify(newCheck));
     handleFilters(newCheck);
   };
 
-  function isChecked(id) {
-    return !(Checked.indexOf(id) === -1);
+
+  const allChecked = (index: number) => {
```
Comment (EnzoVieira) [inline components/CheckBox/CheckBox.tsx:81]: For just maintaining a name convention, for functions that returns a boolean you can use prefixes like 'is', 'has', 'can'. In this case, you can change these name for `isAllChecked` and `isNoneChecked`.

That way, it is easier to others programmers what is his purpose.

File: components/CheckBox/CheckBox.tsx [hunk 3]
```diff
@@ -65,15 +71,45 @@ function CheckBox({ filters, handleFilters }) {
...
+    return !(event[index].some((elem) => { return !Checked.includes(elem.id) }))
+
+  }
+  const noneChecked = (index: number) => {
+    return !event[index].some((elem) => { return Checked.includes(elem.id) })
   }
 
+  const handleToggleAll = (values, index: number, index1: number) => {
```
Comment (EnzoVieira) [inline components/CheckBox/CheckBox.tsx:89]: For a cleaner way to do this function, you can use:

```js
  const handleToggleAll = (values, index: number, index1: number) => {
    let newChecked = [];

    // If all checkboxed are not checked, check them all.
    // Else, newChecked array will continue being empty and will uncheck them all
    if (!isAllChecked(index1 + sync[index])) {
      newChecked = values.map((event) => event.id);
    }
    
    setChecked(newChecked);
    handleFilters(newChecked);
    localStorage.setItem("checked", JSON.stringify(newChecked));
  };
```
That way is more concise and readable.

**Review by EnzoVieira — CHANGES_REQUESTED** [review]
Great job! 🚀

I'd suggest just some changes:

File: components/CheckBox/CheckBox.tsx [hunk 4]
```diff
@@ -65,15 +71,45 @@ function CheckBox({ filters, handleFilters }) {
...
+    else
+      for (const value of values) {
+        if (newCheck.includes(value.id)) {
+          newCheck.splice(newCheck.indexOf(value.id));
+        }
+        setChecked(newCheck);
+        localStorage.setItem("checked", JSON.stringify(newCheck));
+        handleFilters(newCheck);
+      }
+  };
```
Comment (EnzoVieira) [inline components/CheckBox/CheckBox.tsx:110]: also, I'd suggest a easier way by doing this function, such as:

```js
  const handleToggleAll = (values, index: number, index1: number) => {
    let newChecked = [...Checked];

    // If all checkboxed are not checked, check them all.
    // Else, filter the checked events that was not finded in `values`
    if (!isAllChecked(index1 + sync[index])) {
      values.map((event) => newChecked.push(event.id));
    } else {
      newChecked = newChecked.filter(
        (eventId) => !values.find((value) => value.id === eventId)
      );
    }

    setChecked(newChecked);
    handleFilters(newChecked);
    localStorage.setItem("checked", JSON.stringify(newChecked));
  };
```

I think it is a more concise and maintainable way of writing this function.

**Review by EnzoVieira — CHANGES_REQUESTED** [review]

File: components/CheckBox/CheckBox.tsx (hunk 1 above)
Comment (diogogmatos) [reply components/CheckBox/CheckBox.tsx]: True, it's removed ;)

File: components/CheckBox/CheckBox.tsx (hunk 2 above)
Comment (diogogmatos) [reply components/CheckBox/CheckBox.tsx]: Done ;)

File: components/CheckBox/CheckBox.tsx (hunk 4 above)
Comment (diogogmatos) [reply components/CheckBox/CheckBox.tsx]: Thank you!

**Review by EnzoVieira — APPROVED** [review]
Awesome 🚀

Good job guys!

**Pull request merged**